Support for GADTs

Meowcolm024 opened the pull request:
Add support for GADTs

- GADT-reasoning in pattern matching (type annotation needed)
- Typing for type selections
- Parsing the wildcard type `?` in type arguments
- use `as` keyword for upcast (same as `:`)
- test cases `GADT[1-6].mls` in `gadt` directory, and `AsOp.mls`
- refactored wildcard type arguments
- correct extruded type args

Some TODOs (added by @LPTK):
- [x] Merge branch `local-gadt-wip` after we figure out the new errors
- [ ] Merge my local changes to simplify away compositions with `??X` types into main branch before merging this PR

---

The PR has become way too big and it's difficult to keep track of what changes cause what results or regressions. Let's split it up and merge each of these things separately:
- [ ] Typing for type selections
- [ ] Parsing the wildcard type `?` in type arguments, use `as` keyword for upcast (same as `:`)
- [ ] Refactored wildcard type arguments, correct extruded type args
- [ ] Rethink the way GADT reasoning is triggered: only introduce skolems when they correspond to user-annotated pattern variable type arguments (as in `App[f, a]`); _and_ in this case don't try to refine these type arguments in the scrutinee.

LPTK committed:
WIP Start implementing branch-local reasoning

Hit some pp problems due to incorrect use of TypeBounds; need to rm TBs from typer output & use proper wilcards instead
We now type patmat branches inside local polymorphic context; this causes regressions in type simplif and some tests to run out of recursion depth limit

Meowcolm024 committed:
refactor

Meowcolm024 committed:
naive type member lookup

Meowcolm024 committed:
WIP type member partially works

Meowcolm024 committed:
WIP fixed name mangling ing TypeHelpers
TODO add support for raw type

Meowcolm024 committed:
WIP refactor

Meowcolm024 committed:
TODO wildcard type and refinement

Meowcolm024 committed:
added `as` keyword

Meowcolm024 committed:
WIP wildcard types

Meowcolm024 committed:
WIP

Meowcolm024 committed:
WIP

Meowcolm024 committed:
WIP added :GADTs flag

Meowcolm024 committed:
added GADT examples

Meowcolm024 committed:
WIP update examples

Meowcolm024 committed:
added some GADT examples

Meowcolm024 committed:
updated some examples and added some notes

Meowcolm024 committed:
WIP update tests

Meowcolm024 committed:
WIP

LPTK committed:
Add an interesting test

Meowcolm024 committed:
WIP move cache to ctx

Meowcolm024 committed:
Merge branch 'local-gadt' of github.com:Meowcolm024/mlscript into local-gadt

Meowcolm024 committed:
annot to trigger GADTs

Meowcolm024 committed:
clean up tests

Meowcolm024 committed:
Merge remote-tracking branch 'hkust-taco/new-definition-typing' into local-gadt

Meowcolm024 committed:
clean up

Meowcolm024 committed:
fix lift

LPTK commented:
Cool! Will try to take a look later. In the meantime, plz read this again and make sure all the changes are clean: https://github.com/hkust-taco/mlscript/blob/mlscript/CONTRIBUTING.md

Meowcolm024 committed:
add keyword "as" for asc

Meowcolm024 committed:
Merge remote-tracking branch 'hkust-taco/new-definition-typing' into local-gadt

Meowcolm024 committed:
cleanup

Meowcolm024 committed:
add more test cases

Meowcolm024 commented:
Known issue: recursion depth exceeded (infinite extrusion loop) when typing certain terms (see `GADT3.mls` and `GADT5.mls` etc)
Observation: it seems that the loop follows a pattern of: `a''# & S <: b'`

Meowcolm024 committed:
Merge branch 'new-definition-typing' into local-gadt

Meowcolm024 committed:
small test case change

Meowcolm024 committed:
add variance in type members decl

Meowcolm024 committed:
Merge branch 'new-definition-typing' into local-gadt

Meowcolm024 committed:
added more tests

Meowcolm024 committed:
added some multi param type tests

LPTK commented on shared/src/test/diff/nu/AsOp.mls:
```diff
@@ -0,0 +1,54 @@
+//│ res
+//│     = 'hello'
+
+// what ?
```
?

LPTK commented on shared/src/test/diff/parser/IfThenElse.mls:
```diff
@@ -447,13 +447,7 @@ let Some(x) = v
 // TODO support?
```
```suggestion
```

Meowcolm024 committed:
Merge remote-tracking branch 'origin/new-definition-typing' into local-gadt

Meowcolm024 committed:
wip

Meowcolm024 committed:
wip

LPTK commented on shared/src/main/scala/mlscript/syntax.scala:
```diff
@@ -234,6 +235,9 @@ final case class NuFunDef(
 final case class Constructor(params: Tup, body: Blk) extends DesugaredStatement with ConstructorImpl // constructor(...) { ... }
 
 
+final case class TypeParamInfo(varinfo: Option[VarianceInfo], visible: Boolean) {
```
```suggestion
final case class TypeParamInfo(varinfo: Opt[VarianceInfo], visible: Bool) {
```

LPTK commented on shared/src/main/scala/mlscript/syntax.scala:
```diff
@@ -234,6 +235,9 @@ final case class NuFunDef(
 
 
+final case class TypeParamInfo(varinfo: Option[VarianceInfo], visible: Boolean) {
+  def getVarOr(default: VarianceInfo): VarianceInfo = varinfo.getOrElse(default)
```
This method seems entirely redundant.

Meowcolm024 replied:
removed

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -33,10 +33,9 @@ class ConstraintSolver extends NormalForms { self: Typer =>
         : Either[Diagnostic, NuMember]
         = {
-    val info = ctx.tyDefs2.getOrElse(clsNme, ???/*TODO*/)
+    val info = ctx.tyDefs2.getOrElse(clsNme, lastWords(s"Can't find member `$clsNme`"))
```
This is a compiler crash, not a proper compiler error. Please fix or leave a TODO.

LPTK replied:
You misunderstood me. I'm not saying you SHOULD crash. Do you want your compiler to crash when it does not find a type member or do you want it to produce a proper error??

Meowcolm024 replied:
fixed crash

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -33,10 +33,9 @@ class ConstraintSolver extends NormalForms { self: Typer =>
+    val info = ctx.tyDefs2.getOrElse(clsNme, lastWords(s"Can't find member `$clsNme`"))
     
     if (info.isComputing) {
-      
```
Remove whitespace change.

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -657,7 +659,8 @@ class ConstraintSolver extends NormalForms { self: Typer =>
       val originalVars = ty.getVars
-      val res = extrude(ty, lowerLvl, pol, upperLvl)(ctx, MutMap.empty, MutSortMap.empty, reason)
+      val res = extrude(ty, lowerLvl, pol, upperLvl)(ctx, ctx.extrCache, ctx.extrCache2, reason)
+      // val res = extrude(ty, lowerLvl, pol, upperLvl)(ctx, MutMap.empty, MutSortMap.empty, reason)
```
Question: does using this line instead make any difference anywhere? What differences?

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -851,8 +854,8 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
-      case (KEYWORD(":"), l0) :: _ if prec <= outer.prec(':') =>
-        consume
+      case (KEYWORD("as" | ":"), l0) :: _ if prec <= outer.prec(':') =>
+        consume // use both "as" and ":"
```
```suggestion
        consume // * Accept both "as" and ":" for now (maybe remove `:` ascriptions later?)
```

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -1071,15 +1081,17 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
             err(msg"dangling variance information" -> S(loc) :: Nil)
-          case N =>
+          case ((true, S(loc)), _) =>
+            err(msg"dangling visible type member" -> S(loc) :: Nil)
```
We need a test to exercise this.

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -1051,18 +1054,25 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
+        (true, S(l0))
+      case _ => (false, N)
+    }
     val vinfo = yeetSpaces match {
```
Why did you make this a tuple of options? You just made some illegal states representable...

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -1071,15 +1081,17 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
-          case N =>
+          case ((true, S(loc)), _) =>
+            err(msg"dangling visible type member" -> S(loc) :: Nil)
+          case _ =>
```
Don't use a default case. List all possibilities.

LPTK commented on shared/src/main/scala/mlscript/NuTypeDefs.scala:
```diff
@@ -945,7 +946,7 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
     
     lazy val explicitVariances: VarianceStore =
-      MutMap.from(tparams.iterator.map(tp => tp._2 -> tp._3.getOrElse(VarianceInfo.in)))
+      MutMap.from(tparams.iterator.map(tp => tp._2 -> tp._3.getVarOr(VarianceInfo.in)))
```
This is unnecessary. Just use `.varinfo.getOrElse`

LPTK commented on shared/src/main/scala/mlscript/TypeDefs.scala:
```diff
@@ -36,7 +36,7 @@ class TypeDefs extends NuTypeDefs { Typer: Typer =>
     kind: TypeDefKind,
     nme: TypeName,
-    tparamsargs: List[(TypeName, TypeVariable)],
+    tparamsargs: List[(TypeName, TypeVariable)],  // TODO add type member mark
```
What does the TODO mean?

LPTK commented on shared/src/main/scala/mlscript/TypeDefs.scala:
```diff
@@ -343,7 +345,8 @@ class TypeDefs extends NuTypeDefs { Typer: Typer =>
                   val fields = fieldsOf(td.bodyTy, paramTags = true)
                   val tparamTags = td.tparamsargs.map { case (tp, tv) =>
-                    tparamField(td.nme, tp) -> FieldType(Some(tv), tv)(tv.prov) }
+                    // TODO fix type member name mangling
```
What needs to be fixed? Why did you not fix it?

LPTK commented on shared/src/main/scala/mlscript/TypeSimplifier.scala:
```diff
@@ -245,7 +245,7 @@ trait TypeSimplifier { self: Typer =>
                 // * Reconstruct a TypeRef from its current structural components
                 val typeRef = TypeRef(td.nme, td.tparamsargs.zipWithIndex.map { case ((tp, tv), tpidx) =>
-                  val fieldTagNme = tparamField(clsTyNme, tp)
+                  val fieldTagNme = tparamField(clsTyNme, tp, false) // default to `C#A`
```
The use of `false` seems wrong. Just use the correct info here.

Meowcolm024 replied:
I think these are in the logic of using the old defs (`TypeDef`), which doesn't support type members. So it should probably be fine here 🤔

LPTK replied:
Ah, then please add a comment why you pick false here.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -16,7 +16,7 @@ import mlscript.Message._
 class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val newDefs: Bool)
     extends ucs.Desugarer with TypeSimplifier {
-  
+
```
Remove whitespace change.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -1495,32 +1525,44 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
           }
       }
-      val newCtx = ctx.nest
+      val newCtx = ctx.nest // TODO refactor
```
Note: we probably want to refactor this. (I'll do it)

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1150,7 +1150,8 @@ abstract class TyperHelpers { Typer: Typer =>
         if (paramTags) RecordType.mk(td.tparamsargs.map { case (tp, tv) =>
             val tvv = td.getVariancesOrDefault
-            tparamField(defn, tp) -> FieldType(
+            // default to `C#A`
```
Why? Just use the correct value.

LPTK commented on shared/src/test/diff/nu/RawTypes.mls:
```diff
@@ -0,0 +1,121 @@
+//│ res
+//│     = 1
+
+// FIXME
```
Explain why this is a FIXME.

LPTK reviewed (changes requested):
There are large inferred type regressions, for example in `ucs/NestedBranches.mls`, `SimpleRegionDSL_raw.mls`, and `JSON.mls`. Did you figure out why? Could you please give a high-level intuituion?
**EDIT**: also `NuUntyped.mls`, `NuSimplyTyped.mls`

We either need to solve these or add a GADT flag, so these don't just happen by default.

LPTK commented:
So, we need to fix the problems in https://github.com/Meowcolm024/mlscript/tree/local-gadt-wip

Looking at one example: the error you get [here](https://github.com/Meowcolm024/mlscript/blob/9c5e2c5795fbeb3f3c405c57bff58f676bb19b85/shared/src/test/diff/gadt/GADT1.mls#L16-L27) is **correct**:

```scala
abstract class Vec[type L, type T]: Nil[T] | Cons[S, T]
class Nil[T] extends Vec[Z, T]
class Cons[L, T](val h: T, val t: Vec[L, T]) extends Vec[S[L], T]
//│ ╔══[ERROR] Type mismatch in type declaration:
//│ ║  l.18: 	class Cons[L, T](val h: T, val t: Vec[L, T]) extends Vec[S[L], T]
//│ ║        	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── type `L` is not an instance of type `S`
//│ ║  l.18: 	class Cons[L, T](val h: T, val t: Vec[L, T]) extends Vec[S[L], T]
//│ ║        	           ^
//│ ╟── Note: constraint arises from class tag:
//│ ║  l.16: 	abstract class Vec[type L, type T]: Nil[T] | Cons[S, T]
//│ ╙──      	                      
```

`Cons[L, T]` is **indeed** not a `Cons[S, T]` because `L` is not the same type as `S`.

What you really want to write here is `: Nil[T] | Cons[?, T]`. So it seems you need to add support for wildcards.

On the other hand, I think `: Nil | Cons` should *in principle* work, but it currently does not (the map function yields an error) and we could try to find out why.

**EDIT to add:**
For the record, the reason `map` doesn't work with the `: Nil | Cons` self-annotation is that we infer scrutinee type `#Nil | #Cons & { Cons#L = ?L0..?L1; Cons#T = ?T0..?T1 }` with `?T0 <: ?T1 <: ‘A | ~⊥(??T)`, which leads to derived constraint `Vec[A, B] <: {Cons#T}`, which yields an extruded `??T` for the `Cons#T` field, from the `Cons` raw type part of the annotation.
Note: the `⊥(??T)` is really weird; feels like it should be `⊤(??T)` – perhaps a pretty-printing issue — EDIT actually no, a plain old bug https://github.com/hkust-taco/mlscript/commit/abe5e62d891fb3844d96c87135fff64fa1ecc19f
Note: at least in situations like these, we could maybe see that the `Cons#T` refinement in this case could be turned into an equivalent `Vec#T` refinement?

LPTK committed:
WIP Prepare the grounds for SimpleTypeOrWildcard

LPTK commented:
See: https://github.com/LPTK/mlscript/commit/2f193c87bbe7d70c9da1d4cb793ee6471a091416

The goal is to allow:

```scala
abstract class Vec[type L, type T]: Nil[T] | Cons[?, T]
class Nil[T] extends Vec[Z, T]
class Cons[L, T](val h: T, val t: Vec[L, T]) extends Vec[S[L], T]
```

And to elaborate on the previous message, this is what happens if we use `: Nil | Cons`:
* we get constraint `Vec['L, 'A] <: #Nil | #Cons & { Cons#T: ..'A }` for the scrutinee
* ie `#Vec & { L = 'L; T = 'A } & (#Nil | #Cons) <: #Nil | #Cons & { Cons#T: ..'A }`
* leading to constraint `#Cons & { L = 'L; T = 'A } <: #Cons & { Cons#T: ..'A }`
* leading to looking up `Cons#T` in `#Cons`, which yields an extruded skolem not compatible with `'A`

LPTK committed:
Fix bug in extrusion (extruded assigned TV level) and hack around prov accumulation explosion

Meowcolm024 committed:
wip

Meowcolm024 committed:
Merge remote-tracking branch 'lptk/local-gadt-wip' into local-gadt-wip

Meowcolm024 committed:
wip

LPTK committed:
Add current type projection unsoundness example

LPTK commented:
So the way type selections/projections are typed is currently broken, as it falls foul of the very same problem they had in Sala. Here's an example:

```scala
class Wrap[type A]
class Test[type B]
//│ class Wrap[A] {
//│   constructor()
//│ }
//│ class Test[B] {
//│   constructor()
//│ }

fun oops(w: Wrap[Test[Int] & Test[Str]]) = 1 : w.A.B
//│ fun oops: (w: Wrap[in Test[out Int | Str] out Test[in Int | Str out nothing]]) -> nothing

oops(new Wrap)
//│ nothing
//│ res
//│     = 1
```

And here's a Scala 3 issue that's similar: https://github.com/lampepfl/dotty/issues/15578

The root of the issue is being able to use an abstract type that has bad bounds as though it really existed (had a valid instance), whereas we have no evidence that it does. Indeed, the only guarantee we have is that the types of _values_ in the context (and their `val` fields) have good bounds. Types nested in _types_ in the context have no such guarantees.

This means the Vector-tail example is broken. We could solve this by adding a way of doing inversion reasoning, or by some other technique.

Meowcolm024 committed:
WIP wildcard

Meowcolm024 committed:
add more tests

Meowcolm024 committed:
wip: fixed some tests

LPTK committed:
Push ascriptions inside branches for better type checking

The logic needs improvement, as it is not hygienic.
But hygiene will be handled in the pretyper later.

Meowcolm024 committed:
fixed parens

Meowcolm024 committed:
wip err msg

LPTK committed:
Minor: remove TODO

LPTK committed:
Fix freak condition in freshening

LPTK commented on shared/src/test/diff/nu/NuScratch.mls:
```diff
@@ -1,3 +1,302 @@
 :NewDefs
 
+class Ty[type T]
```
We need to remember to clean this file before merging.

LPTK commented on shared/src/main/scala/mlscript/syntax.scala:
```diff
@@ -90,6 +90,7 @@ final case class Forall(params: Ls[TypeVar], body: Term)             extends Ter
 final case class Inst(body: Term)                                    extends Term
 final case class Super()                                             extends Term
 final case class Eqn(lhs: Var, rhs: Term)                            extends Term // equations such as x = y, notably used in constructors; TODO: make lhs a Term
+final case class WildcardType()                                      extends Term // TODO use Var("?") instead
```
Why? What does this do?

Meowcolm024 replied:
It seems that  `WildcardType` actually does nothing, so I decide to remove it (just using `Var("?")` should be fine.

LPTK replied:
But it's still there...

Meowcolm024 replied:
changes not pushed yet

LPTK committed:
Merge branch 'mlscript' into local-gadt (one test now fails)

# Conflicts:
#	compiler/shared/main/scala/mlscript/compiler/ClassLifter.scala
#	shared/src/main/scala/mlscript/ConstraintSolver.scala
#	shared/src/main/scala/mlscript/JSBackend.scala
#	shared/src/test/diff/ecoop23/PolymorphicVariants.mls
#	shared/src/test/diff/nu/Interfaces.mls
#	shared/src/test/diff/nu/PolymorphicVariants_Alt.mls
#	shared/src/test/diff/nu/repro_EvalNegNeg.mls
#	shared/src/test/diff/ucs/HygienicBindings.mls

LPTK committed:
Make test green; not sure why it now fails

Meowcolm024 committed:
wip various fixed

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -648,7 +648,7 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
         exprCont(Super().withLoc(S(l0)), prec, allowNewlines = false)
       case (IDENT("?", true), l0) :: _ => 
```
Why do you need a special case for this?

Meowcolm024 replied:
removing it will cause unexpected operator in parsing

LPTK commented on shared/src/test/diff/gadt/Wildcard.mls:
```diff
@@ -411,42 +420,77 @@ if x is ? then 1
+//│ ╙──       	^^^^^^
+//│ type ? = Int
+
+// TODO should we allow this
```
We should not. `?` is only for class/trait type argument positions.

Meowcolm024 committed:
WIP correct polarity/variance treatment of WildcardArg

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1287,8 +1285,8 @@ abstract class TyperHelpers { Typer: Typer =>
       case WildcardArg(lb, ub) =>
-        if (pol =/= S(true)) apply(S(false))(lb)
-        if (pol =/= S(false)) apply(S(true))(ub)
+         apply(pol.map(!_))(lb); 
```
Why the weird indentation and semicolon? Please try to be more careful.

LPTK committed:
WIP Move towards a correct handling of WildcardArg

LPTK committed:
Add alias GADT use case

Meowcolm024 committed:
various small fixes

Meowcolm024 committed:
change typing of type member alias

Meowcolm024 committed:
implement parser for type argument bounds

LPTK commented on shared/src/main/scala/mlscript/NuTypeDefs.scala:
```diff
@@ -1446,12 +1446,15 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
                     }
                     
-                    val ttu = typeTypingUnit(td.body, S(td))
+                    val ttu = typeTypingUnit(td.body, S(td)) 
```
```suggestion
                    val ttu = typeTypingUnit(td.body, S(td))
```

LPTK commented on shared/src/test/diff/gadt/AliasGADT.mls:
```diff
@@ -5,31 +5,54 @@ type Test[T] = LitInt | LitStr
+//│ ╙──      	        ^
+//│ fun foo: forall 'T. (x: Test['T]) -> 'T
 
+// * somehow we need to find a way to associate T in Foo and T in MkF
```
Right. It should probably be `type Foo[T0] = MkF & { type T = T0 }`.

(Or in the future, using a type refinement syntax, `MkF { T = T0 }`.)

LPTK replied:
Problem is we currently don't support parsing `{ type T = T0 }`.

LPTK replied:
So maybe let's not try to implement this in the current PR. It would require too many additional changes/features.

LPTK commented on shared/src/test/diff/nu/TypeMembers.mls:
```diff
@@ -0,0 +1,64 @@
+//│     = 1
+
+class Foo[T](x: T) { 
+  type T = T
```
Doesn't `T` refer to itself here? 🤔

LPTK commented on shared/src/main/scala/mlscript/NewLexer.scala:
```diff
@@ -442,7 +442,8 @@ object NewLexer {
     "constructor",
-    "virtual"
+    "virtual",
+    "restricts"
```
```suggestion
    "restricts",
```

LPTK commented on shared/src/test/diff/gadt/AliasGADT.mls:
```diff
@@ -5,31 +5,54 @@ type Test[T] = LitInt | LitStr
 class LitStr(s: Str) { type T = Str }
 //│ type Test[T] = LitInt | LitStr
-//│ class LitInt(n: Int) {
-//│   type T = Int
```
TODO: remember to make this pretty-printed again.

Meowcolm024 committed:
small format fix

LPTK commented:
So regarding https://github.com/hkust-taco/mlscript/pull/195#pullrequestreview-1805379488, I minimized the regression to:

```scala
type List[out A] = Cons[A]
module Nil
class Cons[A](head: A, tail: List[A])
//│ type List[A] = Cons[A]
//│ module Nil
//│ class Cons[A](head: A, tail: List[A])

fun listConcat(xs) =
  if xs is
    Cons(x, xs2) then Cons(x, listConcat(xs2))
//│ fun listConcat: forall 'A 'A0 'A1. Cons['A & ('A0 | ~??A)] -> Cons[in 'A0 out 'A1]
//│   where
//│     'A0 :> 'A1 | ??A & 'A
//│     'A1 :> 'A0 | ??A & 'A
```
Previously, we didn't have these duplicated bounds.

I confirmed that they are simply due to extruding things several times. When we either cache the extrusion context or don't increase the level of patmat branches, then we get something like the original inferred type back.

Caching the extrusion context:
```scala
//│ fun listConcat: forall 'A 'A0. Cons['A & ('A0 | ~??A)] -> Cons['A0]
//│   where
//│     'A0 :> ??A & 'A
```

Not increasing the level of patmat branches:
```scala
//│ fun listConcat: forall 'A 'A0. Cons['A] -> Cons['A0]
//│   where
//│     'A0 :> ??A & 'A
```

So we should cache the extrusion context. But to do this properly, we should only cache extrusions _to the same level_, instead of having one single cache that's used for extrusions to any level, as is currently implemented.

Meowcolm024 committed:
WIP added test case similar to the zip one

LPTK commented on shared/src/test/diff/nu/InferredInheritanceTypeArgs.mls:
```diff
@@ -174,12 +174,8 @@ B.foo
 
 
 // * TODO: when 
```
I guess we can remove the TODO.

Meowcolm024 committed:
wip change tag checking in <:< for LhsNf

Meowcolm024 committed:
wip: self type sig issue

LPTK commented on shared/src/main/scala/mlscript/NewParser.scala:
```diff
@@ -1147,13 +1147,13 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
         val lb = getTypeName("restricts")
         val ub = getTypeName("extends")
-
+        // TODO update `TypeParamInfo` to use lb and ub
```
TODO

LPTK commented on shared/src/main/scala/mlscript/NormalForms.scala:
```diff
@@ -241,7 +241,14 @@ class NormalForms extends TyperDatatypes { self: Typer =>
+              case Extruded(true, _) => true
+              case _ => false
+            }
+          } && rt1 <:< rt2 &&
```
This thing looks very fishy. What is going on here? Why don't we use the normal set functions?

Meowcolm024 committed:
WIP early finish some dnfs

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -332,6 +332,21 @@ class ConstraintSolver extends NormalForms { self: Typer =>
+          rhs.cs.exists(c => c.lnf match {
+            case LhsTop => false
+            case LhsRefined(rbs, rtt, rr, rtr) => 
+              !ltt.intersect(rtt).collect{ case sk: SkolemTag => sk }.isEmpty
```
Why not like this? It's much more efficient.
```suggestion
              ltt.exists(rtt.contains)
```
Also I think this could be generalized to all abstract tags, not just skolem tags, no?

Meowcolm024 replied:
maybe not? class tags have type members associated with and need to be compared?

LPTK replied:
A class tag represents the type of a class instance whose type members are unknown. If you find the same one on both sides of a constraint, you can obviously finish the constraint right there.

Meowcolm024 replied:
okay

LPTK commented on shared/src/test/diff/gadt/GADT5.mls:
```diff
@@ -157,22 +157,11 @@ fun g[T](x: F[T, T]): T = if x is
 
 // * We used to be able to manually refine it
 // * but this now fails... not sure why
```
Comment is outdated.

LPTK commented on shared/src/test/diff/gadt/Misc.mls:
```diff
@@ -300,18 +299,75 @@ u(MkU(1))
-//│ ║         	^^^^^^^^^^^^^^^^^
-//│ ╙── expression of type `??S & ~T & ~??S0` does not match type `nothing`
-//│ fun u: forall 'T 'b. (x: U['T, 'b]) -> (error | 'T)
+    MkU(s) then s :x.A
```
```suggestion
    MkU(s) then s : x.A
```

LPTK commented on shared/src/test/diff/gadt/Misc.mls:
```diff
@@ -300,18 +299,75 @@ u(MkU(1))
+//│ abstract class U[A, B]: MkU[anything]
+//│ class MkU[S](x: S) extends U
+
+// TODO no variable leaks now ?
```
Not sure what the TODO mean here. Elaborate?

Meowcolm024 replied:
there was a type variable leak here before we early finish the constrain solving for DNF (when we found skolem at both sides)

LPTK replied:
Ok then we can remove the TODO?

Meowcolm024 committed:
wip change extr type args

Meowcolm024 committed:
add prov for wildcard

Meowcolm024 committed:
WIP refactor wildcard

Meowcolm024 committed:
WIP fix various crashes

Meowcolm024 committed:
WIP fix standalone wildcard typing

Meowcolm024 committed:
wip expand alias issue

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -527,11 +527,11 @@ class ConstraintSolver extends NormalForms { self: Typer =>
             
             case (_, RhsBases(pts, bf, trs)) if trs.nonEmpty =>
-              annoying(Nil, done_ls, trs.valuesIterator.map(_.expand).toList, RhsBases(pts, bf, SortedMap.empty))
+              annoying(Nil, done_ls, trs.valuesIterator.map(_.expand(true)).toList, RhsBases(pts, bf, SortedMap.empty))
```
This should be `false`... The RHS is on the right-hand side, a negative type.

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -992,22 +992,22 @@ class ConstraintSolver extends NormalForms { self: Typer =>
               }
             } else {
-              if (tr1.mayHaveTransitiveSelfType) rec(tr1.expand, tr2.expand, true)
+              if (tr1.mayHaveTransitiveSelfType) rec(tr1.expand(true), tr2.expand(true), true)
```
Why would you use `true` everywhere? It's true/false here...

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -992,22 +992,22 @@ class ConstraintSolver extends NormalForms { self: Typer =>
                   reportError()
                 case _ =>
-                  rec(tr1.expand, tr2.expand, true)
+                  rec(tr1.expand(true), tr2.expand(true), true)
```
Be more careful. true/false again here.

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -992,22 +992,22 @@ class ConstraintSolver extends NormalForms { self: Typer =>
             ()
           case (_, tr: TypeRef) =>
-            rec(lhs, tr.expand, true)
+            rec(lhs, tr.expand(true), true)
```
false

LPTK commented on shared/src/main/scala/mlscript/NuTypeDefs.scala:
```diff
@@ -585,7 +585,7 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
             }
           case td: NuTypeDef =>
-            if (td.nme.name in reservedTypeNames)
+            if (td.nme.name.in(reservedTypeNames) || td.nme.name === "?")
```
Why not put "?" in `reservedTypeNames` instead?

LPTK commented on shared/src/main/scala/mlscript/TypeSimplifier.scala:
```diff
@@ -92,7 +92,7 @@ trait TypeSimplifier { self: Typer =>
       
       case tr @ TypeRef(defn, targs) if builtinTypes.contains(defn) && tr.canExpand =>
-        process(tr.expandOrCrash, parent)
+        process(tr.expandOrCrash(true), parent)
```
This will be incorrect if the type has type parameters.
```suggestion
        assert(tr.targs.isEmpty)
        process(tr.expandOrCrash(true), parent)
```

LPTK commented on shared/src/main/scala/mlscript/TypeSimplifier.scala:
```diff
@@ -268,7 +268,7 @@ trait TypeSimplifier { self: Typer =>
                 println(s"typeRef ${typeRef}")
                 
-                val clsFields = fieldsOf(typeRef.expandWith(paramTags = true, selfTy = false), paramTags = true)
+                val clsFields = fieldsOf(typeRef.expandWith(paramTags = true, selfTy = false, pol.getOrElse(true)), paramTags = true)
```
I don't know how to fix it yet but this is definitely wrong.
```suggestion
                val clsFields = fieldsOf(typeRef.expandWith(paramTags = true, selfTy = false, pol.getOrElse(die/*TODO*/)), paramTags = true)
```

LPTK commented on shared/src/main/scala/mlscript/TypeSimplifier.scala:
```diff
@@ -356,7 +356,7 @@ trait TypeSimplifier { self: Typer =>
                 println(s"typeRef ${typeRef}")
                 
-                val clsFields = fieldsOf(typeRef.expandWith(paramTags = true, selfTy = false), paramTags = true)
+                val clsFields = fieldsOf(typeRef.expandWith(paramTags = true, selfTy = false, pol = pol.getOrElse(true)), paramTags = true)
```
Same.

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -554,9 +554,9 @@ abstract class TyperHelpers { Typer: Typer =>
+          if (primitiveTypes contains tr.defn.name) && tr.canExpand => tr.expandOrCrash(true) <:< that
         case (_, tr: TypeRef)
-          if (primitiveTypes contains tr.defn.name) && tr.canExpand => this <:< tr.expandOrCrash
+          if (primitiveTypes contains tr.defn.name) && tr.canExpand => this <:< tr.expandOrCrash(true)
```
false

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -638,7 +638,7 @@ abstract class TyperHelpers { Typer: Typer =>
     }
     def unwrapAll(implicit ctx: Ctx): SimpleType = unwrapProxies match {
-      case tr: TypeRef if tr.canExpand => tr.expandOrCrash.unwrapAll
+      case tr: TypeRef if tr.canExpand => tr.expandOrCrash(true).unwrapAll
```
Wrong!

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -647,7 +647,7 @@ abstract class TyperHelpers { Typer: Typer =>
       case ComposedType(false, l, r) => l.negNormPos(f, p) | r.negNormPos(f, p)
       case NegType(n) => f(n).withProv(p)
-      case tr: TypeRef if !preserveTypeRefs && tr.canExpand => tr.expandOrCrash.negNormPos(f, p)
+      case tr: TypeRef if !preserveTypeRefs && tr.canExpand => tr.expandOrCrash(true).negNormPos(f, p)
```
Wrong!

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1052,7 +1052,7 @@ abstract class TyperHelpers { Typer: Typer =>
           if (cannotBeDistribbed.isEmpty) S(res)
           else S(PolymorphicType(polymLevel, res))
-        case tr: TypeRef if !traversed.contains(tr.defn) => go(tr.expand, traversed + tr.defn, polymLevel)
+        case tr: TypeRef if !traversed.contains(tr.defn) => go(tr.expand(true), traversed + tr.defn, polymLevel)
```
Wrong!

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1077,8 +1077,8 @@ abstract class TyperHelpers { Typer: Typer =>
         || info.isComputed)
-    // TODO add pol: Bool
-    def expand(implicit ctx: Ctx, raise: Raise): SimpleType = {
+    // TODO provide correct pol at each expand call (we're assuming `pol = true` now)
```
> (we're assuming `pol = true` now)

Just saw that. Really not a good idea as you'll forget to fix most of the places you put the wrong value...

Meowcolm024 replied:
now all fixed...

Meowcolm024 replied:
now all fixed...

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1110,14 +1110,15 @@ abstract class TyperHelpers { Typer: Typer =>
-                // TODO raise type error here
-                SkolemTag(tp._2)(noProv) -> errType
+                // * FIXME we should NOT use TypeBounds here (somehow causes infinite subtyping loop)
+                // *       but raising an error will break a lot of  code related to alias (maybe changling `subst`)
```
This should be an error. We must make sure type aliases never get wildcard args.

Meowcolm024 replied:
it seems that wildcards can get into type aliases via extrusion currently

LPTK replied:
Well extrusion must be changed so this doesn't happen. If the thing is a type alias, its type arguments can use the old extrusion approach (TypeBounds).

Meowcolm024 replied:
done

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1478,7 +1479,7 @@ abstract class TyperHelpers { Typer: Typer =>
           if (!distributeForalls) N else ty match {
         case poly @ PolymorphicType(plvl, bod) => S(poly)
-        case tr: TypeRef if !traversed.contains(tr.defn) => go(tr.expand, traversed + tr.defn)
+        case tr: TypeRef if !traversed.contains(tr.defn) => go(tr.expand(true), traversed + tr.defn)
```
wrong

LPTK commented on shared/src/main/scala/mlscript/TyperHelpers.scala:
```diff
@@ -1500,7 +1501,7 @@ abstract class TyperHelpers { Typer: Typer =>
     def unapply(ty: ST)(implicit ctx: Ctx): S[ST] = {
       def go(ty: ST, traversedVars: Set[TV]): S[ST] = ty match {
-        case tr: TypeRef if tr.canExpand => go(tr.expandOrCrash, traversedVars)
+        case tr: TypeRef if tr.canExpand => go(tr.expandOrCrash(true), traversedVars)
```
wrong

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -521,6 +521,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         }
       case tn @ TypeTag(name) => rec(TypeName(name.decapitalize)) // TODO rm this hack
       // case tn @ TypeTag(name) => rec(TypeName(name))
+      case wc @ TypeName("?") => err(msg"Invalid use of ? here", wc.toLoc)
```
```suggestion
      case wc @ TypeName("?") => err(msg"Invalid use of `?` here", wc.toLoc)
```

Meowcolm024 committed:
wip fix pol

Meowcolm024 committed:
wip fix alias ty arg extursion

Meowcolm024 committed:
wip early finish dnf solving for other tags

LPTK commented on shared/src/main/scala/mlscript/ConstraintSolver.scala:
```diff
@@ -1426,8 +1426,11 @@ class ConstraintSolver extends NormalForms { self: Typer =>
-            // TypeBounds.mk(extrude(targ, lowerLvl, false, upperLvl), extrude(targ, lowerLvl, true, upperLvl)) // Q: ? subtypes?
-            WildcardArg(extrude(targ, lowerLvl, false, upperLvl), extrude(targ, lowerLvl, true, upperLvl))(tr.prov)
+            // * use TypeBounds for type alias
+            if (ctx.tyDefs2.get(d.name).map(_.decl.kind === Als).getOrElse(false))
```
The correct thing to do is to assume we can't youse wildcards, to be on the safe side.
```suggestion
            if (ctx.tyDefs2.get(d.name).map(_.decl.kind === Als).getOrElse(true))
```

Meowcolm024 replied:
Fixed

Meowcolm024 committed:
wip fix extr for alias and updated some tests

Meowcolm024 committed:
small fix

Meowcolm024 committed:
simple level indexed extrusion cache

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -46,6 +46,37 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
+    val cache: MutMap[Level, MutMap[TypeVarOrRigidVar->Bool, TypeVarOrRigidVar]],
+  ) {
+      private var cycle: Int = 0
+      // TODO maybe find a better way
```
Could you explain what this achieves and why it's needed?

LPTK committed:
WIP replace incorrect use of mapTargs and avoid level-mismatched constraint rewritings

LPTK committed:
WW

LPTK committed:
WWW

LPTK committed:
WWW

LPTK committed:
WWW

LPTK committed:
WWW

LPTK committed:
WW

LPTK committed:
WW fix extruded compare

LPTK committed:
W simplify YC's mess

LPTK committed:
WWW trial

LPTK committed:
W

LPTK committed:
W

LPTK committed:
Minor

Meowcolm024 committed:
Merge remote-tracking branch 'origin/mlscript' into local-gadt

Meowcolm024 committed:
add some test cases

Meowcolm024 committed:
fix some test cases

LPTK commented:
Thanks again for the great work, @Meowcolm024! These changes will be useful for future reference, even though they are already obsolete (since we are moving to a new type checker implementation). So I will close this for now as it is not worth fixing the conflicts and cleaning up the history, similarly to https://github.com/hkust-taco/mlscript/pull/219

By the way, as I told you in person some time ago, you merged some commits on my WIP branch (the "W"-named commits) which were only temporary experimentation changes. These were not intended to be made part of your PR – or at least, not as is.
